cpp/src/arrow/util/span.h

```diff
@@ -59,12 +59,13 @@ writing code which would break when it is replaced by std::span.)");
   template <
       typename R,
       typename DisableUnlessConstructibleFromDataAndSize =
-          decltype(span<T>(std::data(std::declval<R>()), std::size(std::declval<R>()))),
+          std::enable_if_t<std::is_convertible_v< std::remove_pointer_t<decltype(std::data(std::declval<R&>()))> (*)[],
+                                              T(*)[]>>,
       typename DisableUnlessSimilarTypes = std::enable_if_t<std::is_same_v<
           std::decay_t<std::remove_pointer_t<decltype(std::data(std::declval<R>()))>>,
           std::decay_t<T>>>>
   // NOLINTNEXTLINE runtime/explicit, non-const reference
-  constexpr span(R&& range) : span{std::data(range), std::size(range)} {}
+  constexpr span(R& range) : span{std::data(range), std::size(range)} {}
```

Comment:
Why not keep the universal reference `R&&`?

Reply:
I ran into a different issue with the universal ref where the template also matched an rvalue input, which is forbidden by one of the tests. I just reverted that change so that you can see it. I believe that the test is correct since you wouldn't want to create a span pointing to a temporary that is immediately discarded. While rvalues can extend the lifetime of objects (see the "rvalue references" section of https://en.cppreference.com/w/cpp/language/reference) that wouldn't make this kind of usage safe since the span would persist after that expression.

Reply:
Ah, fair enough.

Reply:
OK I think I've got a working version of this now. No need to change the reference parameter.